### ISSUE TITLE
Weak reference to the subscriber

### ISSUE THREAD

**Issue by benoitdion** [greenrobot/EventBus#57]
EventBus should keep a weak reference to the subscriber so registration aren't leaking objects when unregister is never called.

**Comment by hoombar** [comment]
+1. Do you have an estimated date this will work its way in?

**Comment by greenrobot** [comment]
No. Probably won't make it into release 3.0.

**Comment by virl** [comment]
Please implement this.

**Comment by virl** [comment]
Please implement this.

**Comment by aprofromindia** [comment]
I would argue - this is just best practice for a pub-sub architecture.

**Comment by guidedways** [comment]
+1

**Comment by guidedways** [comment]
Read this, it's recommended that weak references be used to avoid memory leaks:

https://en.wikipedia.org/wiki/Observer_pattern

This isn't an alien concept. It's expected.

**Comment by techyourchance** [comment]
Hi guys.

IMHO, this is a very desirable feature for anyone who attempts to implement a modular approach, Single Responsibility Principle and a general separation of concerns in their apps. 

Use case: 
I'd like to have some ProductProvider class which is injected into Activities and allows for retrieval of ProductEntity objects. ProductProvider either retrieves products from SQLite cache, or, if the product is not cached yet, initiates background data fetch from server (which is carried out by SyncAdapter). 
In the later case (fetch from server), when background sync completes I need to notify ProductProvider of that event, and EventBus comes in very handy here (much easier than ContentObservers and allows for more "informative" notifications). 

Current approach:
EventBus is injected into ProductProvider, and ProductProvider registers itself when it initiates a background server sync. Since several syncs can be initiated in parallel, ProductProvider keeps track of all the initiated syncs. When all syncs initiated by specific ProductProvider completed, ProductProvider unregisters from EventBus.

Drawbacks of current approach:
1) ProductProvider need to store and manage records of all the syncs it initiated 
2) If any "sync completed" event is not sent (e.g. due to some minor bug), we leaked ProductProvider and all the objects strongly referenced from it
3) Fetch of data from server can take some time, therefore, even if all events are delivered, it might happen that this notification is not relevant anymore. The performance of the app could be better if ProductProvider would be GC'ed earlier.

Alternative approach:
Activity which uses ProductProvider could register/unregister it to/from EventBus in its lifecycle callbacks, but that exposes implementation details of ProductProvider and introduces some "usage convention" which can't be enforced in practice.

This thread is more than two years old, but it looks like this feature is still not supported. Now, I really like EventBus and would gladly deep dive into its code and implement support for WeakReferences myself, but I'd like to be sure that this work, if done, is a desirable feature that will be merged into EventBus in the next release (otherwise I'll have hard time updating to new releases, while  maintaining this manual addition).

So guys, what do you think about this?

**Comment by nateridderman** [comment]
At a minimum, you should add the unregister step to your README. Most users of your library will want to do this.

**Comment by IlyaEremin** [comment]
Any changes here? I catch memory leak when fast add\remove fragment from activity (it can occur when user move back and forth between parts of the app). 
I register fragment in `onResume` and unregister in `onPause`, so `WeakReference` will help a lot.

**Comment by Armenm** [comment]
+1

**Comment by nicjansma** [comment]
This recently caused me issues, not knowing EventBus was holding a strong reference on my Android fragments and other UI classes, eventually leading to a lot of OOMs.

It does look like a note was recently added to the README though! https://github.com/greenrobot/EventBus/commit/8bfdfe8dd2840393b260c93df3bf2bcc664dfe9d

**Comment by guidedways** [comment]
Not trying to turn people away from EventBus, but for this reason and a lot more, I've personally moved to using RxJava. Creating a simple 'bus' (which replaces EventBus entirely) in RxJava is a lot simpler than it sounds (literally 10 lines of code or less), and what you get is a more deterministic behaviour. With RxJava you could still cleanly dispose your consumables but the resultant code is easier to maintain and you get thread safety since you can opt a particular thread to observe and react to fired events.

**Comment by william-ferguson-au** [comment]
Can you provide an example of the 10 lines of codes required to replace EventBus using RxJava?

**Comment by guidedways** [comment]
Sure

Here's my RxBus.java class

```java
public class RxBus {
  public static final RxBus INSTANCE = new RxBus();

  private final Subject<Object> _bus;

  private RxBus() {
    _bus = PublishSubject.create().toSerialized();
  }

  public void post(Object o) {
    _bus.onNext(o);
  }

  public io.reactivex.Observable<Object> asObservable() {
    return _bus.hide();
  }
}
```

You post the same way:

```java
RxBus.INSTANCE.post(new EventSomethingHappened());
```

And you can re-use the existing `onEvent` methods by observing like so:

```java
RxBus.INSTANCE.asObservable()
        .observeOn(AndroidSchedulers.mainThread())
        .subscribe(event -> {
          if (event instanceof EventSomethingHappened) {
            onEventSomethingHappened();
          } else if (event instanceof EventSomethingElse) {
            onEventSomethingElse((EventSomethingElse) event);
          }
        }));
```

**Comment by anhvt52** [comment]
I also met leaking memory problem when register/unregister in onAttachedToWindow /onDetachedFromWindow of custom view because custom view holds the current activity(mContext) so onDetachedFromWindow never called and the Activity will be leaked. WeakReference is the easy way to handle the leak.

**Comment by mimoccc** [comment]
Leaking is a big trouble for this lib.
Same here, i call register and unregister, but there is still leak.
Using a singleton in android app is unusual.
I use AIDL service for event bus, and should be solution for this library too.
In my case i need bus in whole application, therefore i am initialising it in Application context, and have no control when, where and what for is application destroyed, therefore i can not call cleanup for evenbus defailtInstance object.
and unsure about how it handle crash of an application.
